data-processing/entities/sentences/extractor.py

```diff
@@ -37,6 +41,86 @@
+PATTERN_LABEL = r"\\label\{[A-Za-z0-9 \\_.,:-]*\}"
+PATTERN_ABSTRACT_BEGIN = r"\\begin\{abstract\}"
+PATTERN_ABSTRACT_END = r"\\end\{abstract\}"
+PATTERN_TABLE_BEGIN = r"\\begin\{table[*]*\}"
+PATTERN_TABLE_END = r"\\end\{table[*]*\}"
+PATTERN_FIGURE_BEGIN = r"\\begin\{figure[*]*\}"
+PATTERN_FIGURE_END = r"\\end\{figure[*]*\}"
+PATTERN_REF = r"\\ref\{[A-Za-z0-9 \\_.,:-]*\}"
+#TODO @dykang Fix to capture citations patterns like \cite[GRUs;][]{cho2004}
+PATTERN_CITE = r"\\cite[A-Za-z0-9 \\_.,:-]*\{[A-Za-z0-9 \\_.,:-]*\}"
+PATTERN_SYMBOL = r"\$[A-Za-z0-9\\ \{\}\(\)\[\]^&*_.,\+:\;-\=#]*\$"
+PATTERN_ANY = r"\\[A-Za-z0-9\\_.,:-]*[\{[A-Za-z0-9 \\_.,:-]*\}]*"
+
+# objects for detecting nesting structures from the extended tex
+NESTING_CHARACTERS_MAPPING = {'{':'}', '(':')', '[':']'}
+
+
+def check_nesting_structure(nesting_chars):
+    stack = []
+    for nchar in nesting_chars:
+        if nchar in ['{','(', '[']:
+            stack.append(nchar)
+        elif nchar in ['}',')',']']:
+            if len(stack) == 0:
+                return False
+            else:
+                if NESTING_CHARACTERS_MAPPING[stack[-1]] != nchar:
+                    return False
+                else:
+                    stack.pop()
+        else:
+            return False
+    return True
+
+
+def check_sentence_or_not(tex, entity_dict):
+    if tex is None:
+        return False
+
+    # check whether text in section or not
+    if 'current_section' not in entity_dict or not entity_dict['current_section']:
+        return False
+    # check whether text is caption in figure/table or not. currently, we remove all text including captions in figure/table
+    # TODO @dykang later, distinguish lines in table and lines in captions
+    if 'current_figure' in entity_dict and entity_dict['current_figure']:
+        return False
+    if 'current_table' in entity_dict and entity_dict['current_table']:
+        return False
+
+    # check nesting structures from tex_sub
+    nesting_characters_in_tex = []
+    for c in tex:
+        if c in list(NESTING_CHARACTERS_MAPPING.keys()) + list(NESTING_CHARACTERS_MAPPING.values()):
+            nesting_characters_in_tex.append(c)
+    if not nesting_characters_in_tex:
+        return True
+    return check_nesting_structure(nesting_characters_in_tex)
+
+
+
+def extract_richer_tex(context_tex, tex):
+    """
+    if surrounding tex includes multiple matched tex cases, we regard the most right-handed one is the final one. This is not a perfect heuristic though for some cases including next line inside the content
+    """
+    if context_tex is '':
+        return None
+    surrounding_tex = context_tex.split(tex)
+    if len(surrounding_tex) > 2:
+        surrounding_tex = [tex.join(surrounding_tex[:-1]), surrounding_tex[-1]]
+    before_tex, after_tex = surrounding_tex
+    richer_tex = before_tex.split('\n')[-1] + tex + after_tex.split('\n')[0]
+    return richer_tex
+
+
+def extract_text_from_tex_entity(entity):
```

Comment:
Is this a more accurate name? Is this just extracting text from a TeX group or an argument to a command?

```suggestion
def extract_text_from_tex_group(entity):
```

Reply:
The new name sounds better!